### ISSUE TITLE
RSpec can't match errors in jruby-9.4.0.0

### ISSUE THREAD

**Issue by pboling** [jruby/jruby#7112]
**Environment Information**
  
Provide at least:
```
❯ jruby -v
jruby 9.4.0.0-SNAPSHOT (3.1.0) 2022-02-08 3562498c80 Java HotSpot(TM) 64-Bit Server VM 13.0.2+8 on 13.0.2+8 +jit [darwin-x86_64]
❯ uname -a
Darwin SpaceBar.local 21.2.0 Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000 arm64
❯ gem -v
3.3.7
❯ bundler -v
Bundler version 2.3.7
```

Other relevant info you may wish to add:

**Expected Behavior**

RSpec and MiniTest can both match errors that are raised.  In all other tested versions of Ruby, JRuby, TruffleRuby this works.

**Actual Behavior**

Only MiniTest can match the errors.  RSpec cannot. It does not matter if the matching is on string equality or regex pattern matching.

# GREEN TEST - MiniTest POC
```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'hashie'
  gem 'minitest'
end

puts "RUBY_VERSION: #{RUBY_VERSION} - RUBY_ENGINE: #{RUBY_ENGINE} - JRUBY_VERSION: #{JRUBY_VERSION}"

class StrictKeyAccessHash < Hash
  include Hashie::Extensions::StrictKeyAccess # Raise KeyError on failed key access
end

require 'minitest/autorun'

class TestCalculator < Minitest::Test
  def setup
    @hash = StrictKeyAccessHash[foo: "bar"]
  end

  def test_success
    assert_equal "bar", @hash[:foo]
  end

  def test_equality
    exception = assert_raises KeyError do
      @hash["oops"]
    end
    assert_equal("key not found: \"oops\"", exception.message)
  end

  def test_matching
    exception = assert_raises KeyError do
      @hash["flare"]
    end
    assert_match(/flare/, exception.message)
  end
end
```

## Output from MiniTest

```shell
❯ jruby jruby_minitest_key_error_match_working.rb
RUBY_VERSION: 3.1.0 - RUBY_ENGINE: jruby - JRUBY_VERSION: 9.4.0.0-SNAPSHOT
Run options: --seed 30060

# Running:

...

Finished in 0.070110s, 42.7900 runs/s, 85.5800 assertions/s.

3 runs, 6 assertions, 0 failures, 0 errors, 0 skips
jruby jruby_minitest_key_error_match_working.rb  14.45s user 1.79s system 201% cpu 8.046 total
```

# RED TEST - RSpec POC
```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'hashie'
  gem 'rspec'
end

puts "RUBY_VERSION: #{RUBY_VERSION} - RUBY_ENGINE: #{RUBY_ENGINE} - JRUBY_VERSION: #{JRUBY_VERSION}"

class StrictKeyAccessHash < Hash
  include Hashie::Extensions::StrictKeyAccess # Raise KeyError on failed key access
end

require 'rspec/autorun'

RSpec.describe StrictKeyAccessHash do
  before do
    @hash = StrictKeyAccessHash[foo: "bar"]
  end

  it "test_success" do
    expect("bar").to eq @hash[:foo]
  end

  it "matches exact" do
    expect {
      @hash["oops"]
    }.to raise_error(KeyError, "key not found: \"oops\"")
  end

  it "matches regex" do
    expect {
      @hash["flare"]
    }.to raise_error(KeyError, /flare/)
  end
end
```

## Output from RSpec
```shell
❯ jruby jruby_rspec_key_error_match_not_working.rb
RUBY_VERSION: 3.1.0 - RUBY_ENGINE: jruby - JRUBY_VERSION: 9.4.0.0-SNAPSHOT
.FF

Failures:

  1) StrictKeyAccessHash matches exact
     Failure/Error:
       expect {
         @hash["oops"]
       }.to raise_error(KeyError, "key not found: \"oops\"")

       expected KeyError with "key not found: \"oops\"", got #<KeyError: key not found: "oops"> with backtrace:
         # jruby_rspec_key_error_match_not_working.rb:28:in `block in <main>'
         # jruby_rspec_key_error_match_not_working.rb:29:in `block in <main>'
     # jruby_rspec_key_error_match_not_working.rb:29:in `block in <main>'

  2) StrictKeyAccessHash matches regex
     Failure/Error:
       expect {
         @hash["flare"]
       }.to raise_error(KeyError, /flare/)

       expected KeyError with message matching /flare/, got #<KeyError: key not found: "flare"> with backtrace:
         # jruby_rspec_key_error_match_not_working.rb:34:in `block in <main>'
         # jruby_rspec_key_error_match_not_working.rb:35:in `block in <main>'
     # jruby_rspec_key_error_match_not_working.rb:35:in `block in <main>'

Finished in 0.21987 seconds (files took 0.73576 seconds to load)
3 examples, 2 failures

Failed examples:

rspec jruby_rspec_key_error_match_not_working.rb:26 # StrictKeyAccessHash matches exact
rspec jruby_rspec_key_error_match_not_working.rb:32 # StrictKeyAccessHash matches regex
```

**Comment by postmodern** [comment]
Ran into this as well on jruby-9.4.0.0 and rspec 3.11.0:

### Ruby code

```ruby
class Test

  def test
    raise(KeyError,"test")
  end

end
```

### RSpec test

```ruby
require 'spec_helper'

describe Test do
  describe "#test" do
    it "should test" do
      expect {
        subject.test
      }.to raise_error(KeyError,"test")
    end
  end
end
```

### Ruby 3.1.3

```

Test
  #test
    should test

Finished in 0.00258 seconds (files took 0.09602 seconds to load)
1 example, 0 failures
```

### JRuby 9.4.0.0 (3.1.0)

````
Failures:

  1) Test#test should test
     Failure/Error:
       expect {
         subject.test
       }.to raise_error(KeyError,"test")
     
       expected KeyError with "test", got #<KeyError: test> with backtrace:
         # ./lib/test.rb:4:in `test'
         # ./spec/test_spec.rb:10:in `block in <main>'
         # ./spec/test_spec.rb:11:in `block in <main>'
     # ./spec/test_spec.rb:11:in `block in <main>'
```

**Comment by enebo** [comment]
All in one file confirming bug:

```ruby
require 'rspec'

class Test

  def test
    raise(KeyError,"test")
  end

end

describe Test do
  describe "#test" do
    it "should test" do
      expect {
        subject.test
      }.to raise_error(KeyError,"test")
    end
  end
end
```

This will end up being something pretty simple since the error message looks like we are returning the right thing but it isn't matching for some reason.

**Comment by headius** [comment]
Weird that this regressed on 9.4. Investigating.

**Comment by headius** [comment]
This appears to be triggered by `did_you_mean` and likely has to do with the `Correctable` module that gets prepended into exceptions.

https://github.com/ruby/did_you_mean/blob/3e506439359112757d3790a5d5f498384ec7db77/lib/did_you_mean/core_ext/name_error.rb

There's still something not right with `Method#super_method` in the presence of `prepend` but I have not been able to reproduce with a simple script.

**Comment by headius** [comment]
This appears to have been fixed by #7575, which was applied to 9.3 and had not been merged until 030e5129ee.
